cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java

```diff
@@ -0,0 +1,179 @@
+
+  private final TransactionRunner transactionRunner;
+
+  @Inject
+  CredentialIdentityStore(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
+  }
+
+  public Collection<CredentialIdentityId> list(String namespace) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(CredentialProvisionerStore.CREDENTIAL_IDENTITIES);
+      Field<?> namespaceIndex = Fields.stringField(CredentialProvisionerStore.NAMESPACE_FIELD,
+          namespace);
+      try (CloseableIterator<StructuredRow> iterator = table.scan(Range.singleton(
+          Collections.singleton(namespaceIndex)), Integer.MAX_VALUE)) {
+        return identitiesFromRowIterator(iterator);
+      }
+    }, IOException.class);
+  }
+
+  public Optional<CredentialIdentity> get(CredentialIdentityId id) throws IOException {
+    Utils.validateResourceName(id.getName());
+    return TransactionRunners.run(transactionRunner, context -> {
+      Optional<StructuredRow> row = fetchRow(context, id);
+      if (!row.isPresent()) {
+        return Optional.empty();
+      }
+      return Optional.of(GSON.fromJson(row.get()
+          .getString(CredentialProvisionerStore.IDENTITY_DATA_FIELD), CredentialIdentity.class));
+    }, IOException.class);
+  }
+
+  public void create(CredentialIdentityId id, CredentialIdentity identity)
+      throws AlreadyExistsException, IOException {
+    Utils.validateResourceName(id.getName());
+    TransactionRunners.run(transactionRunner, context -> {
+      if (fetchRow(context, id).isPresent()) {
+        throw new AlreadyExistsException(String.format("Credential identity '%s:%s' already exists",
+            id.getNamespace(), id.getName()));
+      }
+      writeIdentity(context, id, identity);
+    }, AlreadyExistsException.class, IOException.class);
+  }
+
+  public void update(CredentialIdentityId id, CredentialIdentity identity)
+      throws IOException, NotFoundException {
+    Utils.validateResourceName(id.getName());
+    TransactionRunners.run(transactionRunner, context -> {
+      if (!fetchRow(context, id).isPresent()) {
+        throw new NotFoundException(String.format("Credential identity '%s:%s' not found",
+            id.getNamespace(), id.getName()));
+      }
+      writeIdentity(context, id, identity);
+    }, IOException.class, NotFoundException.class);
+  }
+
+  public void delete(CredentialIdentityId id) throws IOException, NotFoundException {
+    Utils.validateResourceName(id.getName());
+    TransactionRunners.run(transactionRunner, context -> {
+      if (!fetchRow(context, id).isPresent()) {
+        throw new NotFoundException(String.format("Credential identity '%s:%s' not found",
+            id.getNamespace(), id.getName()));
+      }
+      StructuredTable table = context.getTable(CredentialProvisionerStore.CREDENTIAL_IDENTITIES);
+      Collection<Field<?>> key = Collections.unmodifiableList(Arrays.asList(
```

Comment:
nit: normally unmodifiable collections are used when returning a collection to prevent callers from modifying internal state. It shouldn't be needed if we're just creating keys.

Reply:
Done